hamilton/plugins/jupyter_magic.py

```diff
@@ -14,26 +14,118 @@
 
 """
 
-import json
+import ast
+import importlib
 import os
-import sys
 from pathlib import Path
 from types import ModuleType
+from typing import Any, Dict, List, Literal, Optional, Set, Union
 
 from IPython.core.magic import Magics, cell_magic, line_magic, magics_class
 from IPython.core.magic_arguments import argument, magic_arguments, parse_argstring
+from IPython.core.shellapp import InteractiveShellApp
 from IPython.display import HTML, display
 
-from hamilton import ad_hoc_utils, driver, lifecycle
+from hamilton import ad_hoc_utils, driver
+from hamilton.lifecycle import base as lifecycle_base
 
 
-def create_module(source: str, name: str = None) -> ModuleType:
-    """Create a temporary module from source code"""
-    module_name = ad_hoc_utils._generate_unique_temp_module_name() if name is None else name
-    module_object = ModuleType(module_name, "")
-    sys.modules[module_name] = module_object
-    exec(source, module_object.__dict__)
-    return module_object
+def get_assigned_variables(module_node: ast.Module) -> Set[str]:
+    """Get the set of variable names assigned in a AST Module"""
+    assigned_vars = set()
+
+    def visit_node(ast_node):
+        if isinstance(ast_node, ast.Assign):
+            for target in ast_node.targets:
+                if isinstance(target, ast.Name):
+                    assigned_vars.add(target.id)
+
+        for child_node in ast.iter_child_nodes(ast_node):
+            visit_node(child_node)
+
+    visit_node(module_node)
+    return assigned_vars
+
+
+def get_function_assigned_variables(code):
+    assigned_variables = []
+
+    def visit_Assign(node):
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                assigned_variables.append(target.id)
+
+    tree = ast.parse(code)
+    for node in ast.walk(tree):
+        if isinstance(node, ast.FunctionDef):
+            for sub_node in ast.walk(node):
+                if isinstance(sub_node, ast.Assign):
+                    visit_Assign(sub_node)
+
+    return assigned_variables
+
+
+def execute_and_get_assigned_values(shell: InteractiveShellApp, cell: str) -> Dict[str, Any]:
+    """Execute source code from a cell in the user namespace and collect
+    the values of all assigned variables into a dictionary.
+    """
+    shell.ex(cell)
+    expr = shell.input_transformer_manager.transform_cell(cell)
+    expr_ast = shell.compile.ast_parse(expr)
+    return {name: shell.user_ns[name] for name in get_assigned_variables(expr_ast)}
+
+
+def rebuild_driver(
```

Comment:
Add docstring + notes that this is using internal APIs/we'll want to replace with a more official pattern

Reply:
I don't think we need this function given we're passed a builder.

Reply:
We do. I applied the changes you requested for the user-facing API to pass a Builder instead of a Driver. Then, behind the scene, I do
```python
temp_module = module_from_source(cell)
dr = builder.build()
dr = rebuild_driver(dr, temp_module)   # exact API doesn't matter
```
 Otherwise, working with Builder directly is a mess:
 ```python
 temp_module = module_from_source(cell)
 dr = builder.with_modules(temp_module).build()
 ```
 It leads to a ton of hard to track bug because `builder` is stateful and comes from the user namespace. Therefore, it keeps reference to old copies of `temp_module`, which need to be manually removed and result in many `function foo already defined` errors